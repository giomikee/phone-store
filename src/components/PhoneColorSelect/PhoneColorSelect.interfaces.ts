import type { ColorOption } from '../../interfaces';

export interface Props {
    onSelectedColor: (hexCode: string) => void;
    colorOptions?: ColorOption[];
    selectedColor?: string;
}