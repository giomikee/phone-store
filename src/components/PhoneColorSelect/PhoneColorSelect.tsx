import { FormControl, InputLabel, MenuItem, Select, Skeleton } from '@mui/material';
import type { Props } from './PhoneColorSelect.interfaces';

const PhoneColorSelect = ({ colorOptions, selectedColor, onSelectedColor }: Props) => {
    if (!colorOptions) {
        return <Skeleton height={32} />
    }

    return (
        <FormControl
            fullWidth 
            variant='filled'
        >
            <InputLabel 
                htmlFor='color-select'
                id='color-select-label'
            >
                Color
            </InputLabel>

            <Select
                color={selectedColor ? 'success' : undefined}
                id='color-select'
                data-testid='color-select'
                name='color-select'
                label='Color'
                labelId='color-select-label'
                value={selectedColor ?? ''}
                onChange={(e) => onSelectedColor(e.target.value)}
            >
                {
                    colorOptions.map(({ hexCode, name }) => (
                        <MenuItem 
                            key={hexCode} 
                            value={hexCode}
                            sx={{ color: hexCode }}
                        >
                            {name}
                        </MenuItem>
                    ))
                }
            </Select>
        </FormControl>
    )
}

export default PhoneColorSelect