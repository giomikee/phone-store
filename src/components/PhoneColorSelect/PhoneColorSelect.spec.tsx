import { describe, expect, it, vi } from 'vitest';
import { fireEvent, render, screen } from '@testing-library/react';
import '@testing-library/jest-dom/vitest';
import PhoneColorSelect from './PhoneColorSelect';

describe('Given PhoneColorSelect component', () => { 
    const defaultProps = {
        colorOptions: [
            {
                hexCode: 'hexCode',
                name: 'name',
                imageUrl: 'imageUrl'
            }
        ],
        onSelectedColor: vi.fn(),
    }

    describe.each([
        defaultProps.colorOptions,
        undefined
    ])('When component is rendered with colorOptions=%s', (colorOptions) => {
        it('Then it should match the snapshot', () => {
            const { container } = render(<PhoneColorSelect {...defaultProps} colorOptions={colorOptions} />);
            expect(container).toMatchSnapshot();
        });
    });

    describe('When selected color is changed', () => {
        it('Then onSelectedColor should be called with the selected color', () => {
            render(<PhoneColorSelect {...defaultProps} />);

            const select = screen.getByTestId('color-select').querySelector('[name="color-select"]');

            expect(select).not.toBeNull();
            
            fireEvent.change(select!, { target: { value: 'hexCode' } });
            
            expect(defaultProps.onSelectedColor).toHaveBeenCalledWith('hexCode');
        });
    });
}); 