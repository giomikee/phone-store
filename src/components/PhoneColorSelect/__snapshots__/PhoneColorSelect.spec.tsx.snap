// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Given PhoneColorSelect component > When component is rendered with colorOptions={ hexCode: 'hexCode', name: 'name', imageUrl: 'imageUrl' } > Then it should match the snapshot 1`] = `
<div>
  <div
    class="MuiFormControl-root MuiFormControl-fullWidth css-ytlejw-MuiFormControl-root"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-filled MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-filled css-1u43uf3-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink="false"
      for="color-select"
      id="color-select-label"
    >
      Color
    </label>
    <div
      class="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-colorPrimary MuiInputBase-formControl MuiSelect-root css-uq6ac5-MuiInputBase-root-MuiFilledInput-root-MuiSelect-root"
      data-testid="color-select"
    >
      <div
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-labelledby="color-select-label color-select"
        class="MuiSelect-select MuiSelect-filled MuiInputBase-input MuiFilledInput-input css-1y2qbti-MuiNativeSelect-root-MuiSelect-select-MuiInputBase-input-MuiFilledInput-input"
        id="color-select"
        role="combobox"
        tabindex="0"
      >
        <span
          aria-hidden="true"
          class="notranslate"
        >
          â€‹
        </span>
      </div>
      <input
        aria-hidden="true"
        aria-invalid="false"
        class="MuiSelect-nativeInput css-j0riat-MuiSelect-nativeInput"
        name="color-select"
        tabindex="-1"
        value=""
      />
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconFilled css-1rfqz7c-MuiSvgIcon-root-MuiNativeSelect-root-MuiSelect-icon"
        data-testid="ArrowDropDownIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`Given PhoneColorSelect component > When component is rendered with colorOptions=undefined > Then it should match the snapshot 1`] = `
<div>
  <span
    class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-kkcbhd-MuiSkeleton-root"
    style="height: 32px;"
  />
</div>
`;
